Source/WebGPU/WGSL/GlobalVariableRewriter.cpp

```diff
@@ -0,0 +1,281 @@
+/*
+ * Copyright (c) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "GlobalVariableRewriter.h"
+
+#include "AST.h"
+#include "ASTVisitor.h"
+#include "CallGraph.h"
+#include "WGSL.h"
+#include <wtf/HashMap.h>
+#include <wtf/HashSet.h>
+
+namespace WGSL {
+
+class RewriteGlobalVariables : public AST::Visitor {
```

Comment:
I kind of think we should have a design of the compiler where there are 2 pieces:

1. A "frontend" which is WGSL-specific, and knows nothing about MSL. Its job is to do WGSL-specific tasks (e.g. type checking, l-value checking, etc.).
2. A "backend" which  knows about MSL. The backend is unable to fail; its whole job is to take the program which we've proved to be correct in the frontend, and to spit out MSL code for it.

And I think we should have a deliberate API between the two ends of the compiler, because that would dramatically help us if we ever want to add another backend generator language one day. (And, ideally all the code to the backend of the compiler would even be in a subdirectory named `msl/` or `Metal Shading Language/`, maybe even with a `MSL::` C++ namespace.)

E.g. This global variable rewriter is kind of an implementation detail of MSL; nothing in WGSL itself cares about global variables being in one form or another. So this would be in the MSL-specific portion. If we want to emit another language, we wouldn't need this pass, right?

Reply:
I agree with you, but I don't really know yet how much of this would be useful for other backends, as you know my experience here is "limited". I do agree that at the moment this pass is tailored to MSL, as that's our only output, but the passes are constructed in a way in which they are easily composable, so it should be a matter of moving them and running them conditionally when the time comes. I'm still figuring some things out as I go, so I'm trying to keep things composable, but not worrying too much about the details, as long as it's something simple to refactor in the future. Does that make sense?